File: clients/brpc-client-python/brpc/client/brpc_api.py
'''
Created on 2015-6-7

@author: Administrator
'''
from brpc.client import brpc_client


class BRpcApi(object):
    '''brpc api for python'''
    def __init__(self, url, service, username, passwd, parent=None):
        self.url = url
        self.service_name = service
        self.parent = parent
        self.subs = {}
        if not parent:
            self.client = brpc_client.BRpcClient(self.url, username, passwd)

    def add_sub_service(self, name):
        self.subs[name] = BRpcApi(name, None, None, None, self)

    def remove_sub_service(self, name):
        del self.subs[name]

    def get_sub_service(self, name):
        return self.subs[name]

    def get_client(self):
        if not self.parent:
            return self.client
        return self.parent.get_client()

    def get_service_name(self):
        if self.parent:
            return '%s.%s' % (self.parent.get_service_name(),
                              self.service_name)
        else:
            return self.service_name

    def get_full_ame(self, method):
        return self.get_service_name() + '.' + method

    def call(self, method, args, cast=False):
        ''' invoke method
        @param method: str
            method name to invoke
        @param args: list
            args of method
        '''
        name = self.get_full_ame(method)
        client = self.get_client()
        return client.call(name, args, cast)

    def cast(self, method, args):
        return self.call(method, args, True)

    def ping(self):
        return self.call('ping', [])

    def echo(self, string):
        return self.call('echo', [string])

    def help(self):
        return self.call('help', [])

    def listMethods(self):
        return self.call('listMethods', [])

    def listServices(self):
        return self.call('listServices', [])

    def listVars(self):
        return self.call('listVars', [])

    def signatureOf(self, methodName):
        return self.call('signatureOf', [methodName])

    def typeOfVar(self, varName):
        return self.call('typeOfVar', [varName])

File: clients/brpc-client-java/src/brpc/client/BrpcApi.java
/**
 *
 */
package brpc.client;

import java.util.HashMap;
import java.util.List;

import brpc.common.exceptions.BrpcException;
import brpc.common.exceptions.InvalidArgsException;
import brpc.common.types.ObjectList;

/**
 * Brpc Api
 * @author Administrator
 * @date 2015-8-31
 */
public class BrpcApi {

	private String serviceName;
	private BrpcClient rpcClient;
	private BrpcApi parent;
	private HashMap<String, BrpcApi> subServices;

	private static final ObjectList nullArgs = new ObjectList();

	public BrpcApi(String service, BrpcApi parent)
	{
		this.init(service, parent, null);
	}

	public BrpcApi(String service, String url, String username, String passwd)
	{
		BrpcClient rpcClient = new BrpcClient(url, username, passwd);
		this.init(service, null, rpcClient);
	}

	protected void init(String service, BrpcApi parent, BrpcClient rpcClient)
	{
		this.serviceName = service;
		this.rpcClient = rpcClient;
		this.parent = parent;

		if (this.parent != null) {
			this.rpcClient = this.parent.rpcClient;
			if(!this.parent.addSubSevice(this)) {
				throw new BrpcException("add service failed");
			}
		}
		else if (this.rpcClient == null) {
			throw new InvalidArgsException("both parent and rpcClient are null");
		}
	}

	public void logout()
	{
		this.rpcClient.logout();
	}

	public String ping()
	{
		return this.call("ping", nullArgs);
	}

	public String echo(String str)
	{
		ObjectList args = new ObjectList();
		args.addValue(str);
		return this.call("echo", args);
	}

	public String help()
	{
		return this.call("help", nullArgs);
	}

	public List<String> listMethods()
	{
		return this.call("listMethods", nullArgs);
	}

	public List<String> listVars()
	{
		return this.call("listVars", nullArgs);
	}

	public List<String> listServices()
	{
		return this.call("listServices", nullArgs);
	}

	public List<String> listEextendServices()
	{
		return this.call("listEextendServices", nullArgs);
	}

	public String signatureOf(String method)
	{
		ObjectList args = new ObjectList();
		args.addValue(method);
		return this.call("signatureOf", args);
	}

	public String typeOfVar(String var)
	{
		ObjectList args = new ObjectList();
		args.addValue(var);
		return this.call("typeOfVar", args);
	}

	public boolean subscribe(String event, String method)
	{
		ObjectList args = new ObjectList();
		args.addValue(event);
		args.addValue(method);
		return this.call("subscribe", args);
	}

	public boolean unsubscribe(String event, String method)
	{
		ObjectList args = new ObjectList();
		args.addValue(event);
		args.addValue(method);
		return call("unsubscribe", args);
	}

	public boolean addSubSevice(BrpcApi service)
	{
		return null != subServices.put(service.name(), service);
	}

	public BrpcApi removeSubSevice(String name)
	{
		BrpcApi service = this.subServices.remove(name);
		return service;
	}

	public BrpcApi getSubService(String name)
	{
		return subServices.get(name);
	}

	public String name()
	{
		return this.serviceName;
	}

	public String object()
	{
		if (parent != null)
			return String.format("%s.%s", parent.object(), name());
		else
			return this.name();
	}

	public <T> T call(String method, ObjectList args) {
		return this.call(object(), method, args);
	}

	public <T> T call(String object, String method, ObjectList args) {
		Object result = this.rpcClient.call(object, method, args);
		return (T)result;
	}

	public <T> T cast(String method, ObjectList args) {
		return this.cast(object(), method, args);
	}

	public <T> T cast(String object, String method, ObjectList args) {
		Object result = this.rpcClient.cast(object, method, args);
		return (T)result;
	}

	/*
	public Object call(String method, ObjectList args) {
		return this.call(object(), method, args);
	}

	public Object call(String object, String method, ObjectList args) {
		Object result = this.rpcClient.call(object, method, args);
		return result;
	}

	public Object cast(String method, ObjectList args) {
		return this.cast(object(), method, args);
	}

	public Object cast(String object, String method, ObjectList args) {
		Object result = this.rpcClient.cast(object, method, args);
		return result;
	}*/

}
